File: src/utils/exportSchema.ts
import type {
  AssetFrameState,
  CountryTerritory,
  FrameAnnotations,
  LegacyDrawingsExport,
  PaletteColor,
  ProjectExport,
  ProjectExportV1,
  ProjectExportV2,
  ProjectState,
  TerritoryDrawings,
  TimelineEntry,
} from '../types/project';
import { v4 as uuidv4 } from 'uuid';
import { createEmptyAnnotations, createEmptyAssetState } from '../types/project';
import { clampDisplaySettings, DEFAULT_DISPLAY_SETTINGS } from '../types/displaySettings';
import { recomputeCountryLabels } from './territoryGeometry';

function normalizeCountry(country: CountryTerritory): CountryTerritory {
  if (!country.regionLabels?.length && country.regions.length > 0) {
    return recomputeCountryLabels({
      ...country,
      regionLabels: country.regionLabels ?? [],
    });
  }
  if (!country.regionLabels) {
    return { ...country, regionLabels: [] };
  }
  return country;
}

function normalizePalette(palette: PaletteColor[]): PaletteColor[] {
  return palette.map((p) => ({
    ...p,
    flagFilename: p.flagFilename ?? null,
  }));
}

function assetStateToExport(state: AssetFrameState): ProjectExportV2['assets'][string][number] {
  return {
    drawings: {
      countries: state.annotations.countries,
      cities: state.annotations.cities,
      divisions: state.annotations.divisions,
      suppressedFactionIds: state.annotations.suppressedFactionIds ?? [],
    },
    infoBoard: {
      date: state.info.dateTitle,
      text: state.info.description,
      factionStats: state.info.factionStats,
    },
  };
}

function normalizeDrawings(drawings: TerritoryDrawings | LegacyDrawingsExport): FrameAnnotations {
  if (drawings && Array.isArray(drawings.countries)) {
    const territoryDrawings = drawings as TerritoryDrawings;
    return {
      countries: territoryDrawings.countries.map(normalizeCountry),
      cities: territoryDrawings.cities ?? [],
      divisions: territoryDrawings.divisions ?? [],
      suppressedFactionIds: territoryDrawings.suppressedFactionIds ?? [],
    };
  }
  return createEmptyAnnotations();
}

function exportToAssetState(entry: ProjectExportV2['assets'][string][number]): AssetFrameState {
  const drawings = entry.drawings;
  return {
    annotations: normalizeDrawings(drawings),
    info: {
      dateTitle: entry.infoBoard?.date ?? '',
      description: entry.infoBoard?.text ?? '',
      factionStats: entry.infoBoard?.factionStats ?? [],
    },
  };
}

export function stateToExport(state: ProjectState): ProjectExportV2 {
  const assets: ProjectExportV2['assets'] = {};
  for (const [filename, copies] of Object.entries(state.assets)) {
    assets[filename] = copies.map(assetStateToExport);
  }

  return {
    version: 2,
    projectName: state.projectName,
    exportedAt: new Date().toISOString(),
    palette: normalizePalette(state.palette),
    carryOverLabels: state.carryOverLabels,
    displaySettings: state.displaySettings,
    assets,
    timeline: state.timeline.map((t) => ({ ...t })),
  };
}

export function isV2Export(data: ProjectExport): data is ProjectExportV2 {
  return data.version === 2;
}

export function isV1Export(data: ProjectExport): data is ProjectExportV1 {
  return data.version === 1;
}

export function migrateV1ToAssets(data: ProjectExportV1): {
  assets: Record<string, AssetFrameState[]>;
  timeline: TimelineEntry[];
} {
  const assets: Record<string, AssetFrameState[]> = {};
  const timeline: TimelineEntry[] = [];

  for (const frame of data.frames) {
    const state: AssetFrameState = {
      annotations: {
        countries: (frame.annotations?.countries ?? []).map((c) =>
          normalizeCountry({ ...c, regionLabels: c.regionLabels ?? [] }),
        ),
        cities: frame.annotations?.cities ?? [],
        divisions: frame.annotations?.divisions ?? [],
      },
      info: frame.info,
    };
    if (!assets[frame.filename]) assets[frame.filename] = [];
    const copyIndex = assets[frame.filename].length;
    assets[frame.filename].push(state);
    timeline.push({
      id: uuidv4(),
      filename: frame.filename,
      copyIndex,
    });
  }

  return { assets, timeline };
}

export function importToAssets(data: ProjectExport): {
  assets: Record<string, AssetFrameState[]>;
  timeline: TimelineEntry[];
  projectName: string;
  palette: ProjectExportV2['palette'];
  carryOverLabels: boolean;
  displaySettings: ProjectExportV2['displaySettings'];
} {
  if (isV2Export(data)) {
    const assets: Record<string, AssetFrameState[]> = {};
    for (const [filename, copies] of Object.entries(data.assets)) {
      assets[filename] = copies.map(exportToAssetState);
    }
    return {
      assets,
      timeline: data.timeline.map((t) => ({ ...t })),
      projectName: data.projectName,
      palette: normalizePalette(data.palette),
      carryOverLabels: data.carryOverLabels,
      displaySettings: clampDisplaySettings(data.displaySettings ?? DEFAULT_DISPLAY_SETTINGS),
    };
  }

  const migrated = migrateV1ToAssets(data);
  return {
    ...migrated,
    projectName: 'Imported Project',
    palette: normalizePalette(data.palette),
    carryOverLabels: data.carryOverLabels,
    displaySettings: DEFAULT_DISPLAY_SETTINGS,
  };
}

export function createInitialAssetsFromFiles(
  filenames: string[],
): Record<string, AssetFrameState[]> {
  const assets: Record<string, AssetFrameState[]> = {};
  for (const filename of filenames) {
    assets[filename] = [createEmptyAssetState()];
  }
  return assets;
}

export function createTimelineFromFiles(filenames: string[]): TimelineEntry[] {
  return filenames.map((filename) => ({
    id: uuidv4(),
    filename,
    copyIndex: 0,
  }));
}

File: backend/app/services/export_schema.py
from __future__ import annotations

import uuid
from datetime import datetime, timezone
from typing import Any

from app.models.project import (
    AssetFrameState,
    CountryTerritory,
    FrameAnnotations,
    FrameInfo,
    PaletteColor,
    ProjectBody,
    ProjectExportV2,
    TimelineEntry,
)
from app.services.geometry import recompute_country_labels


def create_empty_frame_info() -> FrameInfo:
    return FrameInfo()


def create_empty_annotations() -> FrameAnnotations:
    return FrameAnnotations(countries=[])


def create_empty_asset_state() -> AssetFrameState:
    return AssetFrameState(annotations=create_empty_annotations(), info=create_empty_frame_info())


def normalize_country(country: CountryTerritory) -> CountryTerritory:
    if not country.regionLabels and country.regions:
        return recompute_country_labels(country.model_copy(update={"regionLabels": []}))
    if not country.regionLabels:
        return country.model_copy(update={"regionLabels": []})
    return country


def asset_state_to_export(state: AssetFrameState) -> dict[str, Any]:
    return {
        "drawings": {"countries": [c.model_dump() for c in state.annotations.countries]},
        "infoBoard": {
            "date": state.info.dateTitle,
            "text": state.info.description,
            "factionStats": [s.model_dump() for s in state.info.factionStats],
        },
    }


def normalize_drawings(drawings: dict[str, Any]) -> FrameAnnotations:
    if drawings and isinstance(drawings.get("countries"), list):
        countries = [CountryTerritory.model_validate(c) for c in drawings["countries"]]
        return FrameAnnotations(countries=[normalize_country(c) for c in countries])
    return create_empty_annotations()


def export_to_asset_state(entry: dict[str, Any]) -> AssetFrameState:
    drawings = entry.get("drawings") or {}
    info_board = entry.get("infoBoard") or {}
    return AssetFrameState(
        annotations=normalize_drawings(drawings if isinstance(drawings, dict) else {}),
        info=FrameInfo(
            dateTitle=info_board.get("date") or "",
            description=info_board.get("text") or "",
            factionStats=info_board.get("factionStats") or [],
        ),
    )


def state_to_export(project: ProjectBody) -> ProjectExportV2:
    assets: dict[str, list[dict[str, Any]]] = {}
    for filename, copies in project.assets.items():
        assets[filename] = [asset_state_to_export(c) for c in copies]
    return ProjectExportV2(
        projectName=project.projectName,
        exportedAt=datetime.now(timezone.utc).isoformat(),
        palette=project.palette,
        carryOverLabels=project.carryOverLabels,
        assets=assets,
        timeline=[t.model_copy() for t in project.timeline],
    )


def migrate_v1_to_assets(data: dict[str, Any]) -> tuple[dict[str, list[AssetFrameState]], list[TimelineEntry]]:
    assets: dict[str, list[AssetFrameState]] = {}
    timeline: list[TimelineEntry] = []
    for frame in data.get("frames") or []:
        countries = frame.get("annotations", {}).get("countries") or []
        state = AssetFrameState(
            annotations=FrameAnnotations(
                countries=[
                    normalize_country(
                        CountryTerritory.model_validate({**c, "regionLabels": c.get("regionLabels") or []})
                    )
                    for c in countries
                ]
            ),
            info=FrameInfo.model_validate(frame.get("info") or {}),
        )
        filename = frame["filename"]
        if filename not in assets:
            assets[filename] = []
        copy_index = len(assets[filename])
        assets[filename].append(state)
        timeline.append(
            TimelineEntry(id=str(uuid.uuid4()), filename=filename, copyIndex=copy_index)
        )
    return assets, timeline


def import_to_project(data: dict[str, Any]) -> ProjectBody:
    version = data.get("version")
    if version == 2:
        assets: dict[str, list[AssetFrameState]] = {}
        for filename, copies in (data.get("assets") or {}).items():
            assets[filename] = [export_to_asset_state(c) for c in copies]
        return ProjectBody(
            projectName=data.get("projectName") or "Imported Project",
            assets=assets,
            timeline=[TimelineEntry.model_validate(t) for t in data.get("timeline") or []],
            palette=[PaletteColor.model_validate(p) for p in data.get("palette") or []],
            carryOverLabels=data.get("carryOverLabels", True),
            currentTimelineIndex=0,
        )
    assets, timeline = migrate_v1_to_assets(data)
    return ProjectBody(
        projectName="Imported Project",
        assets=assets,
        timeline=timeline,
        palette=[PaletteColor.model_validate(p) for p in data.get("palette") or []],
        carryOverLabels=data.get("carryOverLabels", True),
        currentTimelineIndex=0,
    )


def create_initial_assets_from_files(filenames: list[str]) -> dict[str, list[AssetFrameState]]:
    return {name: [create_empty_asset_state()] for name in filenames}


def create_timeline_from_files(filenames: list[str]) -> list[TimelineEntry]:
    return [TimelineEntry(id=str(uuid.uuid4()), filename=name, copyIndex=0) for name in filenames]
